hudi-client/hudi-client-common/src/main/java/org/apache/hudi/table/action/ttl/strategy/KeepByTimeStrategy.java

```diff
@@ -0,0 +1,110 @@
+
+  public KeepByTimeStrategy(HoodieTable hoodieTable, String instantTime) {
+    super(hoodieTable, instantTime);
+    this.ttlInMilis = writeConfig.getPartitionTTLStrategyDaysRetain() * 1000 * 3600 * 24;
+  }
+
+  @Override
+  public List<String> getExpiredPartitionPaths() {
+    Option<HoodieInstant> lastCompletedInstant = hoodieTable.getActiveTimeline().filterCompletedInstants().lastInstant();
+    if (!lastCompletedInstant.isPresent() || ttlInMilis <= 0
+        || !hoodieTable.getMetaClient().getTableConfig().getPartitionFields().isPresent()) {
+      return Collections.emptyList();
+    }
+    List<String> expiredPartitions = getExpiredPartitionsForTimeStrategy(getPartitionPathsForTTL());
+    int limit = writeConfig.getPartitionTTLMaxPartitionsToDelete();
+    LOG.info("Total expired partitions count {}, limit {}", expiredPartitions.size(), limit);
+    return expiredPartitions.stream()
+        .limit(limit) // Avoid a single replace commit too large
+        .collect(Collectors.toList());
+  }
+
+  protected List<String> getExpiredPartitionsForTimeStrategy(List<String> partitionsForTTLManagement) {
+    HoodieTimer timer = HoodieTimer.start();
+    Map<String, Option<String>> lastCommitTimeForPartitions = getLastCommitTimeForPartitions(partitionsForTTLManagement);
+    LOG.info("Collect last commit time for partitions cost {} ms", timer.endTimer());
+    return lastCommitTimeForPartitions.entrySet()
+        .stream()
+        .filter(entry -> entry.getValue().isPresent())
+        .filter(entry -> isPartitionExpired(entry.getValue().get()))
+        .map(Map.Entry::getKey)
+        .collect(Collectors.toList());
+  }
+
+  /**
+   * @param partitionPaths Partitions to collect stats.
+   */
+  private Map<String, Option<String>> getLastCommitTimeForPartitions(List<String> partitionPaths) {
+    int statsParallelism = Math.min(partitionPaths.size(), 200);
+    return hoodieTable.getContext().map(partitionPaths, partitionPath -> {
+      Option<String> partitionLastModifiedTime = hoodieTable.getHoodieView()
+          .getLatestFileSlicesBeforeOrOn(partitionPath, instantTime, true)
+          .map(FileSlice::getBaseInstantTime)
+          .max(Comparator.naturalOrder())
+          .map(Option::ofNullable)
+          .orElse(Option.empty());
+      return Pair.of(partitionPath, partitionLastModifiedTime);
+    }, statsParallelism).stream().collect(Collectors.toMap(Pair::getKey, Pair::getValue));
+  }
+
+  /**
+   * Determines if a partition's reference time has exceeded its time-to-live (TTL).
+   * <p>
+   * This method checks if the current time has passed the TTL threshold based on a
+   * reference time, which could be the creation time or the last commit time of the partition.
+   *
+   * @param referenceTime last commit time or creation time for partition
+   */
+  protected boolean isPartitionExpired(String referenceTime) {
+    String expiredTime = instantTimePlusMillis(fixInstantTimeCompatibility(referenceTime), ttlInMilis);
```

Comment:
Hi, fixInstantTimeCompatibility is called twice.
https://github.com/xicm/hudi/blob/cab0e2bbb9afd73e2d157612e085a724fbc14c01/hudi-common/src/main/java/org/apache/hudi/common/table/timeline/HoodieInstantTimeGenerator.java#L97-L106